api/src/main/java/io/neba/api/resourcemodels/AnnotatedFieldMapper.java

```diff
@@ -0,0 +1,162 @@
+ *
+ * Licensed under the Apache License, Version 2.0 the "License";
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ * http://www.apache.org/licenses/LICENSE-2.0
+
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ **/
+package io.neba.api.resourcemodels;
+
+import org.apache.sling.api.resource.Resource;
+import org.apache.sling.api.resource.ValueMap;
+
+import java.lang.annotation.Annotation;
+import java.lang.reflect.Field;
+import java.util.Map;
+
+/**
+ * OSGi services implementing this interface may customize the mapping of arbitrary fields during
+ * the resource to model mapping of any resource model.
+ * <p>
+ * The implementing service specifies {@link #getAnnotationType() the annotation it is responsible for}
+ * as well as {@link #getFieldType() the type the mapped fields} must be {@link Class#isAssignableFrom(Class) assignable to}.
+ * </p>
+ * <p>
+ * For example, the following service would be responsible for fields annotated with &#64;MyAnnotation
+ * with a type <em>assignable to</em> {@link java.util.Collection}, e.g. List&lt;Resource&gt; or Set&lt;String&gt;.
+ * </p>
+ * <pre>
+ * &#64;Service
+ * &#64;Component(immediate = true)
+ * public class MyFieldMapper&lt;Collection, MyAnnotation&gt; {
+ *     public Class&lt;? extends Collection&gt; getFieldType() {
+ *         return Collection.class;
+ *     }
+ *
+ *     public Class&lt;Annotation&gt; getAnnotationType() {
+ *         return MyAnnotation.class;
+ *     }
+ *
+ *     public Collection map(OngoingMapping&lt;Collection, MyAnnotation&gt; ongoingMapping) {
+ *         ...
+ *     }
+ * }
+ * </pre>
+ * <p>
+ * Custom mappers are always invoked <em>after</em> all of NEBA's standard mappings have occurred, but before
+ * the corresponding value was set on the model's field. They may thus make use of the already resolved
+ * value or choose to provide a different one.
+ * </p>
+ *
+ * <p>
+ *     <strong>It is crucial for a {@link AnnotatedFieldMapper} to always return a value that is assignment compatible
+ *     to the {@link OngoingMapping#getFieldType() field type}, i.e. either of the same or of a more specific type.</strong>
+ *     It is insufficient to return a type compatible to the {@link #map(AnnotatedFieldMapper.OngoingMapping) mapping methods}
+ *     return type declaration. This return type only represents the type any returned value must be compatible to.<br />
+ *     For instance, if a mapper is responsible for {@link java.util.Collection}, it must take care to return the field's actual
+ *     collection type, e.g. {@link java.util.List} or {@link java.util.Set}. Otherwise, an exception will arise.
```

Comment:
In all use cases I can image, the mapper only wants to add elements to the collection. And he doesn't care about the  collection type. So couldn't NEBA provide an instantiated collection of the correct type?